### PR TITLE
Feature/structure search score cache

### PR DESCRIPTION
**Pull request by julianreichinger**
### Your checklist for this pull request
🚨Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [ ] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [ ] Check the commit's or even all commits' message styles matches our requested structure.
- [ ] Check your code additions will fail neither code linting checks nor unit test.

#### Changes
Please list the proposed changes in this pull request.
- Wrapper class for StructureScore instances with a dynamically sized cache for local scores

💔Thank you!

### CODE REVIEW

**Comment by ankurankan** [comment]
@julianreichinger Thanks a lot for the PR. But one of the tests is failing on python 2.7: https://travis-ci.org/pgmpy/pgmpy/jobs/343734794#L845. And it's because python 2.7 doesn't have `unittest.mock`, and for dealing with that we use the mock module: https://mock.readthedocs.io/en/latest/. Could you please fix the tests (can either have a try except in imports or just replace with the mock module) ?

**Comment by julianreichinger** [comment]
@ankurankan Tests are fixed, I replaced the unittest.mock package.

File: pgmpy/estimators/ScoreCache.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    def __init__(self, base_scorer, data, max_size=10000, **kwargs):
+        """
+        A wrapper class for StructureScore instances, which implement a decomposable score,
+        that caches local scores.
+        Based on the global decomposition property of Bayesian networks for decomposable scores.
+
+        Parameters
+        ----------
+        :param base_scorer: Instance of type StructureScore
```
Comment (ankurankan) [inline pgmpy/estimators/ScoreCache.py]: @julianreichinger We use numpy style doctests. So, it should be something like:
```
base_scorer: StructureScore instance
```

**Review by ankurankan — CHANGES_REQUESTED** [review]
@julianreichinger Sorry for the late reply. The PR looks good except for one minor comment that I have put. Also, could you please merge all your commits into a single commit using `git rebase`. That will help us in keeping our git history clean.

**Comment by julianreichinger** [comment]
@ankurankan I made the change you requested and squashed the commits. Actually I have never done that before but it is nice to finally learn my way around Git.

File: pgmpy/estimators/ScoreCache.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+    def __init__(self, base_scorer, data, max_size=10000, **kwargs):
+        """
+        A wrapper class for StructureScore instances, which implement a decomposable score,
+        that caches local scores.
+        Based on the global decomposition property of Bayesian networks for decomposable scores.
+
+        Parameters
+        ----------
+        :param base_scorer: StructureScore instance
```
Comment (ankurankan) [inline pgmpy/estimators/ScoreCache.py]: @julianreichinger This still doesn't seem right. I am sorry for not being clearer in my last comment. You will need to remove the `:param` from the front for numpy style docstring. You can also have a look at it in more detail here: http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html.

**Comment by julianreichinger** [comment]
@ankurankan Sorry, I misunderstood what you meant. I read through the numpy examples and changed the docstrings accordingly. Hopefully everything should be correct now.

**Comment by julianreichinger** [comment]
@ankurankan Any update?

**Comment by ankurankan** [comment]
@julianreichinger Looks good. Merging. Sorry for the delay.

**Review by ankurankan — APPROVED** [review]

**Pull request merged**